File: warp-csharp/WarpCsharp/AiDemo.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

namespace WarpCsharp
{
    public static class AiDemo
    {
        public static async Task TriggerAiDemo()
        {
            var registry = new ApiProviderRegistry();

            string? openaiKey = Environment.GetEnvironmentVariable("OPENAI_API_KEY");
            if (string.IsNullOrEmpty(openaiKey))
            {
                Console.WriteLine("[Info] No OPENAI_API_KEY found. Falling back to DummyProvider for OpenAI.");
                registry.RegisterProvider("openai", new DummyProvider { Name = "OpenAI-Mock" });
            }
            else
            {
                Console.WriteLine("[Info] Found OPENAI_API_KEY. Registering actual OpenAiProvider.");
                registry.RegisterProvider("openai", new OpenAiProvider(openaiKey));
            }

            registry.RegisterModel(new Model { ProviderID = "openai", ModelID = "gpt-4", ContextWindow = 8192, SupportsTools = true });

            registry.RegisterProvider("anthropic", new DummyProvider { Name = "Anthropic-Mock" });
            registry.RegisterModel(new Model { ProviderID = "anthropic", ModelID = "claude-3-opus", ContextWindow = 200000, SupportsTools = true });

            var ctx = new Context
            {
                SystemPrompt = "You are a helpful coding assistant.",
                Messages = new List<Message> { new Message { Role = MessageRole.User, Content = new List<ContentPart> { new ContentPart { Type = ContentPartType.Text, Text = "Write a short haiku about coding." } } } },
                Tools = new List<string>()
            };

            Console.WriteLine("=== Executing AI Stream (OpenAI) ===");
            var stream1 = registry.ExecuteStream("openai", "gpt-4", ctx);
            await foreach (var ev in stream1.ReadAllAsync())
            {
                if (ev.Type == EventType.TextDelta) Console.Write(ev.Data);
                else if (ev.Type == EventType.Error) Console.WriteLine($"\n[Error] {ev.Error?.Message}");
                else if (ev.Type == EventType.Done) Console.WriteLine("\n[Done]");
            }

            Console.WriteLine("=== Executing AI Stream (Anthropic) ===");
            var stream2 = registry.ExecuteStream("anthropic", "claude-3-opus", ctx);
            await foreach (var ev in stream2.ReadAllAsync())
            {
                if (ev.Type == EventType.TextDelta) Console.Write(ev.Data);
                else if (ev.Type == EventType.Error) Console.WriteLine($"\n[Error] {ev.Error?.Message}");
                else if (ev.Type == EventType.Done) Console.WriteLine("\n[Done]");
            }
        }
    }
}

File: warp-java/src/main/java/dev/warp/AiDemo.java
package dev.warp;

public class AiDemo {
    public static void triggerAiDemo() {
        AiProvider.ApiProviderRegistry registry = new AiProvider.ApiProviderRegistry();

        String openaiKey = System.getenv("OPENAI_API_KEY");
        if (openaiKey == null || openaiKey.isEmpty()) {
            System.out.println("[Info] No OPENAI_API_KEY found. Falling back to DummyProvider for OpenAI.");
            AiProvider.DummyProvider openai = new AiProvider.DummyProvider();
            openai.name = "OpenAI-Mock";
            registry.registerProvider("openai", openai);
        } else {
            System.out.println("[Info] Found OPENAI_API_KEY. Registering actual OpenAiProvider.");
            registry.registerProvider("openai", new AiProvider.OpenAiProvider(openaiKey));
        }

        AiProvider.Model gpt4 = new AiProvider.Model();
        gpt4.providerID = "openai";
        gpt4.modelID = "gpt-4";
        gpt4.contextWindow = 8192;
        gpt4.supportsTools = true;
        registry.registerModel(gpt4);

        AiProvider.DummyProvider anthropic = new AiProvider.DummyProvider();
        anthropic.name = "Anthropic-Mock";
        registry.registerProvider("anthropic", anthropic);

        AiProvider.Model claude3 = new AiProvider.Model();
        claude3.providerID = "anthropic";
        claude3.modelID = "claude-3-opus";
        claude3.contextWindow = 200000;
        claude3.supportsTools = true;
        registry.registerModel(claude3);

        AiProvider.Context ctx = new AiProvider.Context();
        ctx.systemPrompt = "You are a helpful coding assistant.";
        AiProvider.Message msg = new AiProvider.Message();
        msg.role = AiProvider.MessageRole.USER;
        AiProvider.ContentPart pt = new AiProvider.ContentPart();
        pt.type = AiProvider.ContentPartType.TEXT;
        pt.text = "Write a short haiku about coding.";
        msg.content.add(pt);
        ctx.messages.add(msg);

        try {
            System.out.println("=== Executing AI Stream (OpenAI) ===");
            java.util.concurrent.Flow.Publisher<AiProvider.AssistantMessageEvent> stream1 = registry.executeStream("openai", "gpt-4", ctx);
            java.util.concurrent.CountDownLatch latch1 = new java.util.concurrent.CountDownLatch(1);
            stream1.subscribe(new java.util.concurrent.Flow.Subscriber<>() {
                public void onSubscribe(java.util.concurrent.Flow.Subscription s) { s.request(Long.MAX_VALUE); }
                public void onNext(AiProvider.AssistantMessageEvent ev) {
                    if (ev.type == AiProvider.EventType.TEXT_DELTA) System.out.print(ev.data);
                    else if (ev.type == AiProvider.EventType.ERROR) System.out.println("\n[Error] " + ev.error);
                    else if (ev.type == AiProvider.EventType.DONE) System.out.println("\n[Done]");
                }
                public void onError(Throwable t) { t.printStackTrace(); latch1.countDown(); }
                public void onComplete() { latch1.countDown(); }
            });
            latch1.await();

            System.out.println("=== Executing AI Stream (Anthropic) ===");
            java.util.concurrent.Flow.Publisher<AiProvider.AssistantMessageEvent> stream2 = registry.executeStream("anthropic", "claude-3-opus", ctx);
            java.util.concurrent.CountDownLatch latch2 = new java.util.concurrent.CountDownLatch(1);
            stream2.subscribe(new java.util.concurrent.Flow.Subscriber<>() {
                public void onSubscribe(java.util.concurrent.Flow.Subscription s) { s.request(Long.MAX_VALUE); }
                public void onNext(AiProvider.AssistantMessageEvent ev) {
                    if (ev.type == AiProvider.EventType.TEXT_DELTA) System.out.print(ev.data);
                    else if (ev.type == AiProvider.EventType.ERROR) System.out.println("\n[Error] " + ev.error);
                    else if (ev.type == AiProvider.EventType.DONE) System.out.println("\n[Done]");
                }
                public void onError(Throwable t) { t.printStackTrace(); latch2.countDown(); }
                public void onComplete() { latch2.countDown(); }
            });
            latch2.await();

        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}
